### PR TITLE
Added Torch NOTMAD

### PR DESCRIPTION
**Pull request by wtt102**
- Located in contextualized/dags/torch_notmad
- Contains demo notebook "torch_nm_demo.ipynb" that compares torch notmad with barebones tf notmad (located in "dags/contextualized/torch_notmad/test_notmad." Original tensorflow version has features not implemented here, such as compatability matrices and low rank, so barebones was used for debugging.
- Comments are provided for the classes and some functions.
- Currently using same dataset as at the end of the "main demo."

NOTE: import statements will need to be updated from (ie, `from callbacks import DAG_loss` -> `from dags.contextualized.torch_notmad.callbacks import DAG_loss`

### CODE REVIEW

File: contextualized/dags/torch_notmad/callbacks.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+        my_dag_loss = torch.mean(DAG_loss(preds, plmodule.alpha, plmodule.rho))
+        
+        if my_dag_loss > self.tol*self.h_old:
+            plmodule.alpha = plmodule.alpha + plmodule.rho*my_dag_loss.item()
+            plmodule.rho = plmodule.rho*1.1
+        self.h_old = my_dag_loss
+
+
+class MyPrintingCallback(Callback):
```
Comment (cnellington) [inline contextualized/dags/torch_notmad/callbacks.py:32]: Seems like development code, remove?

File: contextualized/dags/torch_notmad/simulation.py [hunk 2]
```diff
@@ -0,0 +1,369 @@
+import numpy as np
```
Comment (cnellington) [inline contextualized/dags/torch_notmad/simulation.py:1]: TL;DR: This simulation.py and test_data.py won't live in torch_notmad upon release, so it might fit better in test_notmad while we validate these changes.

We need a tests.py for torch NOTMAD to validate our design choices, and eventually we would like a suite of simulation experiments that allow others to easily replicate results. Right now, simulation.py and test_data.py are used to replicate the old NOTMAD results, which achieves a little bit of both. Ideally we would validate design choices for torch-NOTMAD in isolation (e.g. with simple tests like checking for convergence) and do more interesting simulations in a separate experiments directory.

We will also want to remove test_notmad eventually. I'm ok mergeing test_notmad into dev, but we want to remove test_notmad before merging to main and releasing this change since the TF notmad will be depreciated.

File: contextualized/dags/torch_notmad/test_data.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
+import numpy as np
```
Comment (cnellington) [inline contextualized/dags/torch_notmad/test_data.py:1]: TL;DR: This simulation.py and test_data.py won't live in torch_notmad upon release, so it might fit better in test_notmad while we validate these changes.

We need a tests.py for torch NOTMAD to validate our design choices, and eventually we would like a suite of simulation experiments that allow others to easily replicate results. Right now, simulation.py and test_data.py are used to replicate the old NOTMAD results, which achieves a little bit of both. Ideally we would validate design choices for torch-NOTMAD in isolation (e.g. with simple tests like checking for convergence) and do more interesting simulations in a separate experiments directory.

We will also want to remove test_notmad eventually. I'm ok mergeing test_notmad into dev, but we want to remove test_notmad before merging to main and releasing this change since the TF notmad will be depreciated.

File: contextualized/dags/torch_notmad/torch_notmad.py [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+import pytorch_lightning as pl
+import numpy as np
+from contextualized.functions import identity_link
+torch.set_default_tensor_type(torch.FloatTensor)
+
+#local imports
+from graph_utils import project_to_dag_torch
+from torch_utils import DAG_loss, NOTEARS_loss
+
+class NGAM(nn.Module):
```
Comment (cnellington) [inline contextualized/dags/torch_notmad/torch_notmad.py:13]: Remove and use the modules.py NGAM or MLP as the default encoder.

File: contextualized/dags/torch_notmad/torch_notmad.py [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+        return self.link_fn(ret)
+    
+    def _get_new_weights(self, layer_):
+        with torch.no_grad():
+            return nn.parameter.Parameter(
+                    torch.tensor(np.random.uniform(-0.01, 0.01, size=layer_.weight.shape)).float()
+                )
+    
+
+class Explainer(nn.Module):
```
Comment (cnellington) [inline contextualized/dags/torch_notmad/torch_notmad.py:62]: Remove and use the modules.py Explainer. May need to add some extra functionality to the Explainer to get/set the archetypes.

File: contextualized/dags/torch_notmad/torch_notmad.py [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+    
+    def _archetypes(self):
+        return torch.multiply(self.archs, self.mask)
+    
+
+class NOTMAD_model(pl.LightningModule):
+    """
+    NOTMAD model
+    """
+    def __init__(self,
```
Comment (cnellington) [inline contextualized/dags/torch_notmad/torch_notmad.py:103]: Try to make this consistent with the contextualized.regression.lightning_modules args e.g. opt_lr --> learning_rate, encoder_width + n_encoder_hidden_layers --> encoder_kwargs

File: contextualized/dags/torch_notmad/torch_notmad.py [hunk 7]
```diff
@@ -0,0 +1,286 @@
...
+        self.context_shape = self.datamodule.C_train.shape
+        self.feature_shape = self.datamodule.X_train.shape
+
+        #dag/loss params
+        self.project_distance = 0.1
+        self.archetype_loss_params = archetype_loss_params
+        self.alpha, self.rho, self.use_dynamic_alpha_rho = self._parse_alpha_rho(sample_specific_loss_params)
+
+        #torch params
+        self.opt_lr = opt_lr
```
Comment (cnellington) [inline contextualized/dags/torch_notmad/torch_notmad.py:159]: Needs to be self.learning_rate to enable Torch Lightning's auto_lr_find feature.

File: contextualized/dags/torch_notmad/torch_notmad.py [hunk 8]
```diff
@@ -0,0 +1,286 @@
...
+        self.opt_step = opt_step
+        self.automatic_optimization = auto_opt
```
Comment (cnellington) [inline contextualized/dags/torch_notmad/torch_notmad.py:161]: May be dev code, remove this and associated code?

File: contextualized/dags/torch_notmad/torch_notmad.py [hunk 9]
```diff
@@ -0,0 +1,286 @@
...
+        #layer params
+        self.init_mat = init_mat
+
+        #layers
+        if self.encoder_type == "NGAM":
+            self.encoder = NGAM(encoder_input_shape[0], encoder_output_shape[0],
+                                n_hidden_layers=n_encoder_hidden_layers, 
+                                width=encoder_width)
+        else:
+            self.encoder = self._build_encoder(encoder_input_shape[0],
```
Comment (cnellington) [inline contextualized/dags/torch_notmad/torch_notmad.py:178]: _build_encoder is depreciated, should specify encoders from modules.py and allow users to pass encoder_kwargs. See regression.metamodels.py for examples of how this might work.

**Review by cnellington — CHANGES_REQUESTED** [review]
Looks good so far! Overall needs some structural changes to start preparing it for a release, and should try to integrate with existing tools (e.g. contextualized.modules) and match existing patterns (e.g. contextualized.regression.lightning_modules and metamodels)

File: contextualized/dags/torch_notmad/torch_notmad.py (hunk 4 above)
Comment (wtt102) [reply contextualized/dags/torch_notmad/torch_notmad.py:13]: Would it be good to keep this as is until we update the explainer and ngam of modules.py?

**Review by blengerich — CHANGES_REQUESTED** [review]
Thanks @wtt102 for the Torch version of NOTMAD! This PR should be requesting to be merged into `dev` branch rather than `main`. You might need to make a new PR to change that.

File: contextualized/dags/torch_notmad/torch_notmad.py (hunk 5 above)
Comment (cnellington) [reply contextualized/dags/torch_notmad/torch_notmad.py:62]: I just updated the explained to include archetype getters and setters:
https://github.com/cnellington/Contextualized/blob/e5fe0a07fe18b6d4df843dc01820cb3b3087b494/contextualized/modules.py#L39

Last missing piece is the mask used by NOTMAD.  I'd suggest either including the mask in NOTMAD or creating a subclass of the modules.py Explainer that extends the functionality.

File: contextualized/modules.py [hunk 10]
```diff
@@ -4,6 +4,7 @@
 from contextualized.functions import identity_link, identity
 
 
+
```
Comment (cnellington) [inline contextualized/modules.py:7]: whitespace

**Review by cnellington — CHANGES_REQUESTED** [review]
Recently merged modules.py and included some functions that should be used here.

**Comment by blengerich** [comment]
Closing this PR because we will merge into the `dev` branch.
